### ISSUE TITLE
Get hold of Backward graph from a C++ module.

### ISSUE THREAD

**Issue by dinkdeep** [pytorch/pytorch#35215]
Hi Guys
@huzecong  @ezyang @gqchen @albanD 
Is there a way we could get hold of Backward Graph in C++ code .

We have a custom FPGA and asic & we would like to perform training on a multi-chips HW asics, they are similar to nvidia GPU's and we would like to interconnect them the way nvidia DGZ servers do  and want to run training on them, For the inference part we are able  use a Glow backend for Pytorch. This backend doesn't seems to have enough support for training as of now and in particular Gradient Accumlation and exchange across devices and loss functions.
and Pytorch seems to have an extensive support for CUDA.

 (https://github.com/pytorch/glow/blob/master/torch_glow/src/Registration.cpp)
This backend has a python binding and gets hold of Pytorch JIT IR for the forward graph and converts the JIT Graph Nodes to its own internal representation of the Glow nodes and it performs its own differentiate() to generate a differential graph which kind of contains both the forward graph and backward graph (integrated together) but it does not has the way to provide access to backward graph .

To me seems to me that pytorch has different forward graph and a backward graphs and it looks pretty elaborate with many optimizers.

So essentially could there be a way to get hold of backward graph from a C++ module (from glow backend iteslf ) so that I can generate the ONNX/glow nodes from it and create a Graph which can be run on our HW. 
Looks like the Back Graph does not has a JIT IR , Essentially I am looking for some sample code , where I can visit the Nodes and edges of the Graph but I need some more concrete example.
. Although I did see that there seem to be some test cases cpp folder of   pytorch/test/cpp/jit/test_autodiff.cpp  api/autograd.cpp  which can be used .

If you guys can get me some pointer and direction to the sample code that would help a lot.

Looks like there is way to iterate over backward nodes from python using 
 https://github.com/szagoruyko/pytorchviz 

![pytroch_mlp_backward_graph](https://user-images.githubusercontent.com/61411073/77338968-1b898f80-6d51-11ea-8c6d-ad43a8087738.png)

I was reading the article of ezyang from http://blog.ezyang.com/2019/05/pytorch-internals/
and found that we  need something on lines of attached below pic where multi devices are communicating using scatter all reduce. 
![eyzang_grad_calc_5](https://user-images.githubusercontent.com/61411073/77339374-a8cce400-6d51-11ea-9c59-ef3a2af3f942.png)

![data_parrallel_pytorch](https://user-images.githubusercontent.com/61411073/77339856-6d7ee500-6d52-11ea-8551-36a729266171.png)

Essentially there should be some way to route torch.nn.DataParallel() and torch.nn.parallel.DistributedDataParallel() API's to Glow module to get access to gloo/nccl C++ modules inside pytorch way it gets access to JIT IR.

**Comment by albanD** [comment]
Hi,

The IR graph from the jit and the graph made of Nodes for the autograd are two independent things.
In particular, the autograd graph only contains Nodes that can execute arbitrary operations and only represent the backward pass.

**Comment by ezyang** [comment]
If you're tracing, trace the forward and backward and you'll get a usable graph you can export. Our symbolic differentiation support is very modest and unlikely to be enough.

**Comment by dinkdeep** [comment]
Hi @albanD, Looks like to me that the backgraph nodes are generated probably by overloading the Forward graph nodes using automatic differentiation during JIT, It would be great if you can point me to the file and functions in torch I would like to take a look once.  

Hi @ezyang I have seen the trace called from python code on forward graph but not sure for backward graph, BTW would it possible to invoke trace from inside C++ code , because we don't want to tinker the user Python codes(generally they might of of customers ) ,so anything beneath is OK , Can you kindly point me to some sample source snippet or some link , BTW regarding "symbolic differentiation" which piece of code gets exercised through this path, please provide a pointer ti it too.

**Comment by albanD** [comment]
> probably by overloading the Forward graph nodes

It doesn't have to as it just runs the regular pytorch cpp ops. And these ops support autograd. So the backward graph is built automatically independently of the jit.

**Comment by dinkdeep** [comment]
Hi @albanD, @ezyang, ,
Hi Alban, So as I understand the autograd/backward ops are basically the "grad" Variable of the  associated tensor of the user specified Ops and are generated dynamically as per the philosophy of torch which is a dynamic graph execution.

Probably the  JIT IR provides the static model, where the entire graph can be serialised and exported to a file and can be loaded from a c++ module offline . In our current model we try to save the JITed pytroch pt file and load it from the Glow.

The question is does  the JIT contains the Nodes and Variables for backward graph as @ezyang mentioned .

Hi @zdevito @suo 
I am tracing and saving the jit file after backward() call is completed in the JIT do expect to see the 
Nodes for backward graph and how do I get the access to it . I have access to torch::jit::LowerGraph() function
which looks like returns  the torch::jit::Graph graph object and a vector of Parameters std::vector<at::Tensor> &parameters, 

Can this parameters be used in some way, or is there an function similar to  LowerGraph() which can return the GraphRoot, leaf nodes for BG. In might need some sample reference & ptr also.

I see this register_hook( ) function does that helps but that looks more from dynamic mode when user is running a python script .

    data = torch.randn(1,3,16,16)
    target = torch.randn( 1,8, 14,1 )
    output = model(data)
    loss_fn = nn.MSELoss()
    loss = loss_fn(output, target)
    loss.backward()
    module = torch.jit.trace(model,data)
    torch.jit.save(module,"fwd_jit.pt")

**Comment by makslevental** [comment]
@ezyang 

> Our symbolic differentiation support is very modest and unlikely to be enough.

I'm quite late to this party but I'm wondering what this means specifically - is there really a symbolic differentiation engine somewhere inside PyTorch? In particular, I'm curious about this comment on your internals blog post

>though PyTorch JIT does know how to do symbolic differentiation

<mind blown>

Is there a doc on how this works? I guess at minimum can you give me a code pointer?

**Comment by ezyang** [comment]
Yes. Code lives at torch/csrc/jit/runtime/autodiff.cpp

**Comment by Linux-cpp-lisp** [comment]
@ezyang can you elaborate on what you mean by "trace the backward"? What does that look like in code?

For example, this simple module can be scripted:
```python
import torch
class M(torch.nn.Module):
    def forward(self, x):
        x.requires_grad_(True)
        y = x**2
        return torch.autograd.grad([y.sum()], [x])[0]

m = M()
m(3*torch.ones(3))  # =>  tensor([6., 6., 6.])
ms = torch.jit.script(m)
ms(3*torch.ones(3))  # => tensor([6., 6., 6.])
```
but `ms.graph` contains a call to the autograd engine:
```
%18 : Tensor?[] = aten::grad(%11, %13, %9, %9, %16, %16) # <ipython-input-1-1f2f9991b0ce>:6:15
```

How would I go about tracing the actual backwards pass into TorchScript? Thanks :smiley:

**Comment by ezyang** [comment]
cc @asuhan

**Comment by faisalasad3** [comment]
@ezyang I have the same question as @Linux-cpp-lisp. Is there actually a way to trace the backward pass?

Alternatively, if I have a forward function implemented, is there a way to trace the derivative of the forward function with respect to it's inputs?

**Comment by asuhan** [comment]
I vaguely remember an autodiff pass in TorchScript which we used in the early days of pytorch/xla, before moving to the lazy tensors approach. We landed a few changes to core PyTorch to add enough symbolic autodiff support to fully handle ResNet-50.

We stitched the forward and backward pass manually, using the `xla::Call` to call the XLA computation generated for the backward graph. We got the backward graph by calling `differentiate` on the `Graph`. In case you're curious, here's how we used to do it in the pre-lazy tensor days: https://github.com/pytorch/xla/blob/2492ddc516b73eda4299fa11d3a04b29dcd84046/torch_xla/csrc/module.cpp#L392

I'm not sure any of the interfaces we used back then still exist in today's TorchScript.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
